rfcs/20200612-stream-executor-c-api.md

```diff
@@ -336,94 +401,145 @@ typedef struct SP_StreamExecutor {
                            TF_Status* status);
 
   // Causes the host code to synchronously wait for the event to complete.
-  void (*block_host_for_event)(
-      SP_Device* executor, SP_Event event, TF_Status* status);
+  void (*block_host_for_event)(const SP_Device* device, SP_Event event,
+                               TF_Status* status);
+
+  // [Optional]
+  // Causes the host code to synchronously wait for operations entrained onto
+  // stream to complete. Effectively a join on the asynchronous device
+  // operations enqueued on the stream before this program point.
+  // If not set, then corresponding functionality will be implemented
+  // by registering an event on the `stream` and waiting for it using
+  // `block_host_for_event`.
+  void (*block_host_until_done)(const SP_Device* device, SP_Stream stream,
+                                TF_Status* status);
 
   // Synchronizes all activity occurring in the StreamExecutor's context (most
   // likely a whole device).
-  TF_BOOL (*synchronize_all_activity)(SP_Device* executor);
-
-  // Obtains metadata about the underlying device.
-  void (*fill_device_description)(SP_Device* executor,
-                                  SP_DeviceDescription* description,
-                                  TF_Status* status);
+  void (*synchronize_all_activity)(const SP_Device* device, TF_Status* status);
 
   // Enqueues on a stream a user-specified function to be run on the host.
-  TF_BOOL (*host_callback)(SP_Device* executor, SP_Stream* stream,
-                           TF_StatusCallbackFn callback_fn, void* ctx);
+  // `callback_arg` should be passed as the first argument to `callback_fn`.
+  TF_Bool (*host_callback)(SP_Device* device, SP_Stream stream,
+                           SE_StatusCallbackFn callback_fn, void* callback_arg);
 } SP_StreamExecutor;
 
-#define SP_STREAMEXECUTOR_STRUCT_SIZE TF_OFFSET_OF_END(SP_StreamExecutor, host_callback)
+#define SP_STREAMEXECUTOR_STRUCT_SIZE \
+  TF_OFFSET_OF_END(SP_StreamExecutor, host_callback)
+
+typedef struct SE_CreateStreamExecutorParams {
+  size_t struct_size;
+  void* ext;  // reserved for future use
+
+  SP_StreamExecutor* stream_executor;  // output, to be filled by plugin
+} SE_CreateStreamExecutorParams;
+
+#define SE_CREATE_STREAM_EXECUTOR_PARAMS_STRUCT_SIZE \
+  TF_OFFSET_OF_END(SE_CreateStreamExecutorParams, stream_executor)
 
 typedef struct SP_Platform {
   size_t struct_size;
-  
-  // Free form data set by plugin.
-  void* ext;
-  
-  // Platform name
+
+  void* ext;  // free-form data set by plugin
```

Comment:
```suggestion
  void* ext;  // reserved for future use
```

Reply:
`SP_Platform` is populated by the plugin. Therefore, `void* ext` is meant to be a field that plugin can set to anything they want. It is not reserved for future use.

Reply:
I'm resolving this one without making changes then. Please feel free to unresolve it if the change should still be made.